### DISCUSSION TITLE
Meson relinks shared libraries and drops linker settings

### DISCUSSION THREAD

**Discussion by oscarbenjamin** [mesonbuild/meson#13586] [General]
I'm not sure if this is a meson bug or if I am just not understanding how this works properly.

Following on from gh-13046 I have the following in meson.build:
```meson
# Add rpaths for a local build of flint found via pkgconfig
#   https://github.com/mesonbuild/meson/issues/13046
if get_option('add_flint_rpath')
  flint_lib_dir = flint_dep.get_pkgconfig_variable('libdir')
  add_project_link_arguments(
    '-Wl,-rpath=' + flint_lib_dir,
    language: 'c',
  )
endif
```
The idea is that I set `-Dadd_flint_rpath=true` and then meson will set the rpath entries in the shared libraries so that they can find my local build of `libflint.so` when I run the tests in a development checkout on Linux.

This works fine but fails mysteriously when meson automatically reconfigures itself. Specifically I run the following commands:
```
meson setup --wipe build --pkg-config-path=.local/lib/pkgconfig -Dadd_flint_rpath=true
meson compile -C build
spin test
```
This sets up a nice new build directory, sets pkgconfig to find libflint.so and tells meson to install the rpaths cia the linker args. Then `spin test` runs the tests and everything works. If I edit the code and return `meson compile` or `spin test` etc I get a nice incremental rebuild and the tests run.

If after the above command I just run meson compile again then I get this:
```console
$ meson compile -C build
INFO: autodetecting backend as ninja
INFO: calculating backend command to run: /home/oscar/.pyenv/versions/python-flint-3.12/bin/ninja -C /home/oscar/current/active/python-flint/build
ninja: Entering directory `/home/oscar/current/active/python-flint/build'
ninja: no work to do.
```
However if I touch the `meson.build` file then something else happens:
```console
$ touch meson.build
$ meson compile -C build
INFO: autodetecting backend as ninja
INFO: calculating backend command to run: /home/oscar/.pyenv/versions/python-flint-3.12/bin/ninja -C /home/oscar/current/active/python-flint/build
ninja: Entering directory `/home/oscar/current/active/python-flint/build'
[0/1] Regenerating build files.
The Meson build system
Version: 1.4.0
Source dir: /home/oscar/current/active/python-flint
Build dir: /home/oscar/current/active/python-flint/build
Build type: native build
Project name: python-flint
Project version: undefined
C compiler for the host machine: cc (gcc 11.4.0 "cc (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0")
C linker for the host machine: cc ld.bfd 2.38
Cython compiler for the host machine: cython (cython 3.1.0)
Host machine cpu family: x86_64
Host machine cpu: x86_64
Program python3 found: YES (/home/oscar/.pyenv/versions/3.12.0/envs/python-flint-3.12/bin/python3.12)
Dependency gmp found: YES 6.3.0 (cached)
Dependency mpfr found: YES 4.1.0 (cached)
Dependency flint found: YES 3.1.0 (cached)
Build targets in project: 38

python-flint undefined

  User defined options
    pkg_config_path: .local/lib/pkgconfig
    prefix         : /usr
    add_flint_rpath: true

Found ninja-1.11.1.git.kitware.jobserver-1 at /home/oscar/.pyenv/versions/python-flint-3.12/bin/ninja
Cleaning... 0 files.
[38/38] Linking target src/flint/types/acb_theta.cpython-312-x86_64-linux-gnu.so
```
Notice that it didn't just reconfigure. It also built 38 targets with the linker.

Now `spin test` fails because the rpaths are not set any more:
```console
$ spin test
...
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/oscar/current/active/python-flint/build-install/usr/lib/python3.12/site-packages/flint/__init__.py", line 1, in <module>
    from .pyflint import *
  File "flint/pyflint.pyx", line 1, in init flint.pyflint
  File "flint/flint_base/flint_base.pyx", line 1, in init flint.flint_base.flint_base
ImportError: libflint.so.19: cannot open shared object file: No such file or directory
```
I don't quite understand what happens after `touch meson.build`. It seems meson observes the timestamp on the file and decides to reconfigure itself which I guess is fine but then it decides to build 38 targets even though nothing has changed. A full build has 114 targets which is because there are 38 extension modules and each has 3 build steps:
```
.pyx -> .c -> .o -> .so
```
The 38 targets that got rebuilt are all the linking step `.o -> .so`.

So for some reason meson decided to relink everything even though nothing changed. And it also didn't use my linker args properly somehow. Now if I touch meson.build again then I get:
```
$ touch meson.build
$ meson compile -C build -v
...
Found ninja-1.11.1.git.kitware.jobserver-1 at /home/oscar/.pyenv/versions/python-flint-3.12/bin/ninja
Cleaning... 0 files.
ninja: no work to do.
```
So this time it does not think that anything needs to be linked. However the linked files are still broken and don't have the rpath. The only way I have found to get it working again from this state is with `--wipe` (or `rm -r build`).

When meson compiles it uses these arguments in the linking step:
```
First time:
cc  -o src/flint/functions/showgood.cpython-312-x86_64-linux-gnu.so src/flint/functions/showgood.cpython-312-x86_64-linux-gnu.so.p/meson-generated_src_flint_functions_showgood.pyx.c.o -Wl,--as-needed -Wl,--allow-shlib-undefined -shared -fPIC -Wl,-rpath=/home/oscar/current/active/python-flint/.local/lib '-Wl,-rpath,$ORIGIN/../../../../.local/lib' -Wl,-rpath-link,/home/oscar/current/active/python-flint/build/../.local/lib -Wl,--start-group /home/oscar/current/active/python-flint/.local/lib/libgmp.so /home/oscar/current/active/python-flint/.local/lib/libmpfr.so /home/oscar/current/active/python-flint/.local/lib/libflint.so -Wl,--end-group
Second time (after touch meson.build):
cc  -o src/flint/functions/showgood.cpython-312-x86_64-linux-gnu.so src/flint/functions/showgood.cpython-312-x86_64-linux-gnu.so.p/meson-generated_src_flint_functions_showgood.pyx.c.o -Wl,--as-needed -Wl,--allow-shlib-undefined -shared -fPIC -Wl,-rpath= '-Wl,-rpath,$ORIGIN/../../../../.local/lib' -Wl,-rpath-link,/home/oscar/current/active/python-flint/build/../.local/lib -Wl,--start-group /home/oscar/current/active/python-flint/.local/lib/libgmp.so /home/oscar/current/active/python-flint/.local/lib/libmpfr.so /home/oscar/current/active/python-flint/.local/lib/libflint.so -Wl,--end-group
```
So the command lines are not exactly the same. The first one has these:
```
-Wl,-rpath=/home/oscar/current/active/python-flint/.local/lib 
'-Wl,-rpath,$ORIGIN/../../../../.local/lib'
-Wl,-rpath-link,/home/oscar/current/active/python-flint/build/../.local/lib
```
Whereas the second one has
```
-Wl,-rpath= '-Wl,-rpath,$ORIGIN/../../../../.local/lib'
-Wl,-rpath-link,/home/oscar/current/active/python-flint/build/../.local/lib
```
The argument that I told it to add is missing on the second run. I'm not sure what the other arguments are for given that they don't actually result in the rpath being set. More generally I don't understand why the arguments would just be the same in both cases.

Is this a meson bug?

Maybe meson gets confused because one path has `build/..` which after normalisation makes it equivalent to the other path?

Is there a better way of getting the rpaths set? The problem is really just that `meson install` strips out the rpaths during install and I don't want it to do that.

**Comment by eli-schwartz** [comment]
> I don't quite understand what happens after `touch meson.build`. It seems meson observes the timestamp on the file and decides to reconfigure itself which I guess is fine but then it decides to build 38 targets even though nothing has changed. A full build has 114 targets which is because there are 38 extension modules and each has 3 build steps:

Reconfiguring is a logical and reasonable outcome of "updating the meson.build files" -- if you have added additional targets, for example, those need to be calculated and added into the build. They may for example be mandatory in order to successfully re-compile targets whose source *.c files got updated too.

Building 38 targets even though nothing has changed is also a logical and reasonable outcome of ninja believing that something *has*, in fact, changed.

Something has to have changed, because ninja calculates whether to rebuild the targets by checking whether the link command has changed. So if the link command changes, the target is rebuilt *with the new link command.

...

What this question therefore boils down to is, why after running `meson --internal regenerate sourcedir/ builddir/` (that is what ninja does when it sees the meson.build file has been updated), do your rpath arguments disappear? This suggests somehow that meson thinks the project option "add_flint_rpath" is false when reconfiguring. But ordinarily, project options persist during a reconfigure.

**Reply by oscarbenjamin** [reply]
> This suggests somehow that meson thinks the project option "add_flint_rpath" is false when reconfiguring.

The configure output shows it as true:
```
  User defined options
    pkg_config_path: .local/lib/pkgconfig
    prefix         : /usr
    add_flint_rpath: true

Found ninja-1.11.1.git.kitware.jobserver-1 at /home/oscar/.pyenv/versions/python-flint-3.12/bin/ninja
Cleaning... 0 files.
[38/38] Linking target src/flint/types/acb_theta.cpython-312-x86_64-linux-gnu.so
```
